Fix query string truncation while dispatching to QE

hughcapet opened the pull request:
buildGpQueryString function, which prepares query text for sliced plan
dispatch to QE, used not to take into account character length. Thus, if
a query text contained multi-byte characters and a truncated version of
the query text was needed, the function tended to cut a character in the
position defined by QUERY_STRING_TRUNCATE_SIZE variable and produce
incorrect symbols subsequently written to segment logs. Such incorrect
symbols made it impossible to complete queries like select * from
gp_toolkit.__gp_log_segment_ext and raised the following error: ERROR:
invalid byte sequence for encoding "UTF8": 0xd022…

hughcapet committed:
Fix query string truncation for dispatching to QE

buildGpQueryString function, which prepares query text for sliced plan
dispatch to QE, used not to take into account character length. Thus, if
a query text contained multi-byte characters and a truncated version of
the query text was needed, the function tended to cut a character in the
position defined by QUERY_STRING_TRUNCATE_SIZE variable and produce
incorrect symbols subsequently written to segment logs. Such incorrect
symbols made it impossible to complete queries like select * from
gp_toolkit.__gp_log_segment_ext and raised the following error: ERROR:
invalid byte sequence for encoding "UTF8": 0xd022…

darthunix commented on src/backend/cdb/dispatcher/cdbdisp_query.c:
```diff
@@ -871,6 +872,8 @@ buildGpQueryString(DispatchCommandQueryParms *pQueryParms,
 	int			total_query_len;
 	char	   *shared_query,
 			   *pos;
+	int			cnt = 0,
```
May be it is more clear to define `cnt` variable here, but initialize it with 0 near the added loop.

darthunix reviewed (commented):
Great work!
I can suggest to move tests from `input/dispatch.source`/`output/dispatch.source` to `sql/dispatch_encoding.sql`/`expected/dispatch_encoding.out`.

hughcapet committed:
Add regression test

This test verifies the fix made in 66d4975. It executes a query longer
than QUERY_STRING_TRUNCATE_SIZE and containing 2-byte symbol in the cut
site. Segment logs don't have broken byte sequence, so the following
query should complete successfully

hughcapet commented:
Removed query with very-very... long column from the test to get rid of NOTICE phrase in output and the column name truncation. Now it verifies execution of the query with multiple columns. Also moved the test to sql/dispatch_encoding.sql and /expected/dispatch_encoding.out as it was suggested
